File: 905-sort-array-by-parity/905-sort-array-by-parity.py
class Solution:
    def sortArrayByParity(self, nums: List[int]) -> List[int]:
        
        # O(n) Time and O(1) extra space
#         odds = []
#         evens = []
        
#         for n in nums:
#             if n % 2 == 0:
#                 evens.append(n)
#             else:
#                 odds.append(n)
#         i = 0
#         for n in evens:
#             nums[i] = n
#             i += 1
#         for n in odds:
#             nums[i] = n
#             i += 1
#         return nums

        
        # O(n) Time and O(1) extra space
        
        i, j = 0, len(nums) - 1
        
        while i < j:
            while i < len(nums) and nums[i] % 2 == 0: i += 1
            while j >= 0 and nums[j] % 2 != 0: j -= 1
            
            if i < j:
                nums[i], nums[j] = nums[j], nums[i]
        return nums

File: 905-sort-array-by-parity/905-sort-array-by-parity.cpp
class Solution {
public:
    vector<int> sortArrayByParity(vector<int>& nums) {
        return solve(nums);
        return bruteforce(nums);
    }
private:
    vector<int> bruteforce(vector<int>& nums) {
        vector<int> odds;
        vector<int> evens;
        
        for(int n: nums) {
            if(n % 2 == 0) evens.push_back(n);
            else odds.push_back(n);
        }
        int i = 0;
        for(int n: evens)
            nums[i++] = n;
        
        for(int n: odds)
            nums[i++] = n;
        return nums;
    }
    
    vector<int> solve(vector<int>& nums) {
        int i = 0, j = nums.size()-1;
        
        while(i < j) {
            while(i < nums.size() && nums[i] % 2 == 0) ++i;
            while(j > 0 && nums[j] % 2 != 0) --j;
            
            if(i < j) swap(nums[i], nums[j]);
        }
        return nums;
    }
};